### PR TITLE
fixing the call to vim.fn.system

### LINKED ISSUE #1547: Opening Directory in NVIMTREE causes ENAMETOOLONG error
### Description

When opening a directory within the nvimtree window using "o" . It still opens the directory and displays the content, however i do get the following annoying error message. I digged a little into it see https://github.com/neovim/neovim/issues/19914 for more

```Could not start the fs_event watcher for path
/home/bob/Programming/Lab-Tool/.git : ENAMETOOLONG
```
Any idea why this is happening and how I might be able to resolve this?

### Neovim version

```text
NVIM v0.8.0-dev+399-g58323b1fe
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-11 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=malloc -Wsuggest-attribute=cold -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az40-824
```

### Operating system and version

OS: Arch Linux x86_64 Host: Surface Book 3 124I:00037T Kernel: 5.18.16-arch1-1-surface Uptime: 2 days, 4 hours, 58 mins Packages: 1634 (pacman) Shell: fish 3.5.1 Resolution: 3240x2160 WM: awesome Theme: Adwaita [GTK2/3] Icons: Adwaita [GTK2/3] Terminal: alacritty

### nvim-tree version

81eb718

### Minimal config

```lua
-- these line in git/utils.lua are causing the issue i am not shure how to resolve it. The problem is that system takes everything from standard out, but this isn't the desired behavior in this case. If somebody prints for example everytime a new terminal is opened the time this would fail hence a other function is needed see https://github.com/neovim/neovim/issues/19914 for more 
local cmd = "git -C " .. vim.fn.shellescape(cwd) .. " rev-parse --show-toplevel"

local toplevel = vim.fn.system(cmd)
```

### Steps to reproduce

1: When opening a directory within the nvimtree window using "o" . 
2: It still opens the directory and displays the content, however i do get the following annoying error message.

### Expected behavior

No error

### Actual behavior

error

### PR DESCRIPTION
**Pull request by Xyhlon**
it is not desired to call vim.fn.system("whole command"), hence it takes all the output of the shell, there might be undesired info be parsed into the string aswell. see #1547 for more info.
better way of calling vim.fn.system({"whole","command"}) only this way the desired output of the command can be guarantied.

resolves #1547

### CODE REVIEW

**Comment by alex-courtis** [comment]
Related: https://github.com/kyazdani42/nvim-tree.lua/issues/1484

Prior art neotree: https://github.com/nvim-neo-tree/neo-tree.nvim/commit/b30a418275e748a6833996a9c655bb107aae2034

File: lua/nvim-tree/git/utils.lua [hunk 1]
```diff
@@ -9,7 +9,7 @@ function M.get_toplevel(cwd)
   local ps = log.profile_start("git toplevel %s", cwd)
   log.line("git", cmd)
```
Comment (alex-courtis) [inline lua/nvim-tree/git/utils.lua:10]: cmd is unused.

Please replace cmd with `{ "git", "-C", vim.fn.shellescape(cwd), "rev-parse", "--show-toplevel" }`, log it and and pass it to `vim.fn.system(`

File: lua/nvim-tree/git/utils.lua [hunk 2]
```diff
@@ -9,7 +9,7 @@ function M.get_toplevel(cwd)
...
 
-  local toplevel = vim.fn.system(cmd)
+  local toplevel = vim.fn.system { "git", "-C", vim.fn.shellescape(cwd), "rev-parse", "--show-toplevel" }
```
Comment (alex-courtis) [inline lua/nvim-tree/git/utils.lua:12]: This fails:
```
[2022-08-27 13:15:55] [profile] START git toplevel /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon
[2022-08-27 13:15:55] [git] git -C '/home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon' rev-parse --show-toplevel
fatal: cannot change to ''/home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon'': No such file or directory
[2022-08-27 13:15:55] [profile] END   git toplevel /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon  2ms
```

master:
```
[2022-08-27 13:15:28] [profile] START git toplevel /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon
[2022-08-27 13:15:28] [git] git -C '/home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon' rev-parse --show-toplevel
/home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon
[2022-08-27 13:15:28] [profile] END   git toplevel /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon  8ms
[2022-08-27 13:15:28] [profile] START git untracked /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon
[2022-08-27 13:15:28] [git] git -C /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon config status.showUntrackedFiles
[2022-08-27 13:15:28] [profile] END   git untracked /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon  7ms
[2022-08-27 13:15:28] [profile] START git job /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon nil
[2022-08-27 13:15:28] [git] running job with timeout 400ms
[2022-08-27 13:15:28] [git] git --no-optional-locks status --porcelain=v1 --ignored=matching -u
!! doc/tags
[2022-08-27 13:15:28] [git] done
[2022-08-27 13:15:28] [profile] END   git job /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon nil  4ms
[2022-08-27 13:15:28] [git] job success    /home/alex/.local/share/nvim/vundle/nvim-tree.lua.Xyhlon nil
```

```lua
log = {
      enable = true,
      truncate = true,
      types = {
        git = true,
        profile = true,
      },
    }
```

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Thanks for looking at this one.

Needs work.

Fix should also be applied to `should_show_untracked`

**Comment by alex-courtis** [comment]
/nudge @Xyhlon 

This will fix issues that many users are experiencing.

**Review by Xyhlon — COMMENTED** [review]
if you remove vim.fn.shellescape from cwd in the cmd everything works without the fatal error message

**Comment by alex-courtis** [comment]
> if you remove vim.fn.shellescape from cwd in the cmd everything works without the fatal error message

That sounds reasonable. It seems that cwd is being passed as an argument rather than to the command line, so no need to escape.

To validate and test:
* a variety of cwd that include spaces and a variety of special characters e.g. `#` that would normally need to be escaped
* MacOS
* Windows WSL, powershell etc. (no developer available)

**Comment by alex-courtis** [comment]
I've merged master and applied the fix to toplevel, as well as untracked.

I cannot push to your repo. Please grant me access or merge master and apply 
[0001-fix-1547-pass-git-toplevel-cwd-unescaped-pass-git-un.patch.txt](https://github.com/kyazdani42/nvim-tree.lua/files/9621372/0001-fix-1547-pass-git-toplevel-cwd-unescaped-pass-git-un.patch.txt)

**Comment by alex-courtis** [comment]
I'm tempted to change the "msys2 git support" path, however the escaping seems fragile and I have no way to test it.

**Comment by alex-courtis** [comment]
This should "resolve"
#1585
#1586 
#1484 
https://github.com/kyazdani42/nvim-tree.lua/issues/549#issuecomment-1127394585
and others

**Comment by Xyhlon** [comment]
applied patch and access granted

**Comment by alex-courtis** [comment]
Merged and pushed to your fork.

Tested a variety of cases with paths that we would normally escape; everything working as expected without `vim.fn.shellescape`.

Merging PR now.

**Comment by alex-courtis** [comment]
Many thanks for driving this change; this has been problematic for some time.

**Pull request merged**